src/manage_sql_secinfo.c

```diff
@@ -2656,6 +2657,485 @@ insert_cve_from_entry (element_t entry, element_t last_modified,
   return 0;
 }
 
+/**
+ * @brief Get the string value for a specified key from a JSON object.
+ *
+ * @param[in]  object  JSON object
+ * @param[in]  key     The key of the string in the JSON object.
+ *
+ * @return The string out of the JSON object with key "key", if any.
+ *         NULL otherwise.
+ */
+static char*
+json_object_item_string (cJSON *object, char *key)
+{
+  cJSON *value_json;
+
+  value_json = cJSON_GetObjectItemCaseSensitive(object, key);
+  if (cJSON_IsString(value_json))
+    return value_json->valuestring;
+  return NULL;
+}
+
+/**
+ * @brief Get the double value for a specified key from a JSON object.
+ *
+ * @param[in]  object    JSON object
+ * @param[in]  key       The key of the double value in the JSON object.
+ * @param[in]  fallback  The fallback value if the double value is not
+ *                       available.
+ *
+ * @return The double value out of the JSON object with key "key", if any.
+ *         The fallback value otherwise.
+ */
+static double
+json_object_item_double (cJSON *object, char *key, double fallback)
+{
+  cJSON *value_json;
+
+  value_json = cJSON_GetObjectItemCaseSensitive(object, key);
+  if (cJSON_IsNumber(value_json))
+    return value_json->valuedouble;
+  return fallback;
+}
+
+/**
+ * @brief Save the node of a cve match rule tree.
+ *
+ * @param[in]  parent_id  The parent_id of the node. If this value is 0,
+ *                        this node is the root of the tree.
+ * @param[in]  cve_id     The id of the CVE to which the tree belongs.
+ * @param[in]  operator   The operator for the match rules.
+ *
+ * @return The (database) id of the node.
+ */
+static int
+save_node (long int parent_id, long int cve_id, char * operator)
+{
+  long int id;
+
+  id = sql_int64_0
+         ("INSERT INTO scap2.cpe_match_nodes"
+          " (parent_id, cve_id, operator)"
+          " VALUES"
+          " (%i, %i, '%s')"
+          " RETURNING scap2.cpe_match_nodes.id;",
+          parent_id,
+          cve_id,
+          operator);
+
+  return id;
```

Comment:
I think `sql_int64_0` returns `long long int` so may be using `resource_t` makes sense as the return and parameters type for `parent_id` and `cve_id` ?

I also suggest skipping the local variable
```suggestion
  return sql_int64_0
         ("INSERT INTO scap2.cpe_match_nodes"
          " (parent_id, cve_id, operator)"
          " VALUES"
          " (%i, %i, '%s')"
          " RETURNING scap2.cpe_match_nodes.id;",
          parent_id,
          cve_id,
          operator);
```

Reply:
I think the function parameters (`parent_id` and `cve_id`) and return type could also be `resource_t` . Same for `load_nodes` since they are all connected.